nn/parallel/parallel.go

```diff
@@ -0,0 +1,38 @@
+package parallel
+
+// #cgo CFLAGS: -I ${SRCDIR}/../../ -I ${SRCDIR}../../cgotorch/libtorch/include
+// #cgo LDFLAGS: -L ${SRCDIR}/../../cgotorch -Wl,-rpath ${SRCDIR}/../../cgotorch -lcgotorch
+// #cgo LDFLAGS: -L ${SRCDIR}/../../cgotorch/libtorch/lib -Wl,-rpath ${SRCDIR}/../../cgotorch/libtorch/lib -lc10 -ltorch -ltorch_cpu
+// #include "cgotorch/cgotorch.h"
+// Tensor goModuleForward(char *m, Tensor input);
+import "C"
+import (
+	"reflect"
+	"unsafe"
+
+	torch "github.com/wangkuiyi/gotorch"
+	"github.com/wangkuiyi/gotorch/nn"
+)
+
+//export goModuleForward
+func goModuleForward(m *C.char, input C.Tensor) C.Tensor {
+	module := (*(*nn.IModule)(unsafe.Pointer(m)))
+	forward := reflect.ValueOf(module).MethodByName("Forward")
+	args := []reflect.Value{reflect.ValueOf(torch.Tensor{(*unsafe.Pointer)(&input)})}
+	return C.Tensor(forward.Call(args)[0].Interface().(torch.Tensor).T)
+}
+
+// DataParallel Evaluates module(input) in parallel across the given devices.
+// If `devices` is not supplied, the invocation is parallelized across all available CUDA devices.
+// If `outputDevice` is supplied, the final, combined tensor will be placed on this device. If not, it defaults to the first device in devices.
+// In detail, this method performs the following four distinct steps:
+//    1. Scatter the input to the given devices,
+//    2. Replicate (deep clone) the model on each device,
+//    3. Evaluate each module with its input on its device,
+//    4. Gather the outputs of each replica into a single output tensor, located on the `outputDevice`.
```

Comment:
There are two approaches for data parallelism in for multi-GPU training:

- Single-Process Multi-GPU
- Per Process Per GPU

PyTorch [DistributedDataParallel](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L461) has proved that Per Process Per GPU is more efficient.

> Single-Process Multi-GPU is not the recommended mode for "
                "DDP. In this mode, each DDP instance operates on multiple "
                "devices and creates multiple module replicas within one "
                "process. The overhead of scatter/gather and GIL contention "
                "in every forward pass can slow down training. "
                "Please consider using one DDP instance per device or per "
                "module replica by explicitly setting device_ids or "
                "CUDA_VISIBLE_DEVICES.

Reply:
So,  scatter--> parallel apply --> gather is not suggested. Instead, we launch a training process for each device. Each training process does dataloading/forward/backward/allreduce/update individually.